pxtpy/converter.ts

```diff
@@ -609,20 +619,39 @@ namespace pxt.py {
                 if (c0.extendsTypes && c0.extendsTypes.length) {
                     // then t0 is narrowed to t1 if and only if t1 is in the extends relationship
                     if (c0.extendsTypes.some(e => e === c1.qName)) {
-                        return
+                        return true
                     }
                 }
             }
         }
-        // otherwise, we unify
+
+        return false
+    }
+
+    function narrow(e: Expr, constrainingType: Type): void {
+        const t0 = find(typeOf(e))
+        const t1 = find(constrainingType)
+
+        if (isAssignable(t0, t1)) {
+            return;
+        }
+
+        // if we don't know if two types are assignable, we can try to unify them ssin common cases
         // TODO: unification is too strict but should always be sound
-        if (currIteration > 1) {
+        if (isFunType(t0) || isFunType(t1)
+            || isPrimativeType(t0) || isPrimativeType(t1)) {
             unify(e, typeOf(e), t1)
+        } else {
+            // if we're not sure about assinability or unification, we do nothing as future
+            // iterations may unify or make assinability clear.
+            // TODO: Ideally what we should do is track a "constraining type" similiar to how we track .union
+            // per type, and ensure the constraints are met as we unify or narrow types. The difficulty is that
+            // this depends on a more accurate assignability check which will take some work to get right.
```

Comment:
this is where `unifyOrConstrain` would be called, right? maybe leave it as a no-op call to keep track for future?

Reply:
Woops, unifyOrConstrain is dead code I removed